src/VisualStudio/CSharp/Impl/ProjectSystemShim/CSharpProjectShim.OptionsProcessor.cs

```diff
@@ -0,0 +1,289 @@
+                    .WithOptimizationLevel(GetBooleanOption(CompilerOptions.OPTID_OPTIMIZATIONS) ? OptimizationLevel.Release : OptimizationLevel.Debug)
+                    .WithOutputKind(_outputKind)
+                    .WithPlatform(platform)
+                    .WithSpecificDiagnosticOptions(diagnosticOptions)
+                    .WithWarningLevel(warningLevel);
+            }
+
+            private static string GetIdForErrorCode(int errorCode)
+            {
+                return "CS" + errorCode.ToString("0000");
+            }
+
+            private IEnumerable<string> ParseWarningCodes(CompilerOptions compilerOptions)
+            {
+                Contract.ThrowIfFalse(
+                    compilerOptions == CompilerOptions.OPTID_NOWARNLIST ||
+                    compilerOptions == CompilerOptions.OPTID_WARNASERRORLIST ||
+                    compilerOptions == CompilerOptions.OPTID_WARNNOTASERRORLIST);
+
+                foreach (var warning in GetStringOption(compilerOptions, defaultValue: "").Split(new[] { ' ', ',', ';' }, StringSplitOptions.RemoveEmptyEntries))
+                {
+                    var warningStringID = warning;
+                    if (int.TryParse(warning, out var warningId))
+                    {
+                        warningStringID = GetIdForErrorCode(warningId);
+                    }
+
+                    yield return warningStringID;
+                }
+            }
+
+            private bool? GetNullableBooleanOption(CompilerOptions optionID)
+            {
+                return (bool?)_options[(int)optionID];
+            }
+
+            private bool GetBooleanOption(CompilerOptions optionID)
+            {
+                return GetNullableBooleanOption(optionID).GetValueOrDefault(defaultValue: false);
+            }
+
+            private string GetFilePathRelativeOption(CompilerOptions optionID)
+            {
+                var path = GetStringOption(optionID, defaultValue: null);
+
+                if (string.IsNullOrEmpty(path))
+                {
+                    return null;
+                }
+
+                var directory = Path.GetDirectoryName(_visualStudioProject.FilePath);
+
+                if (!string.IsNullOrEmpty(directory))
+                {
+                    return FileUtilities.ResolveRelativePath(path, directory);
+                }
+
+                return null;
+            }
+
+            private string GetStringOption(CompilerOptions optionID, string defaultValue)
+            {
+                string value = (string)_options[(int)optionID];
+
+                if (string.IsNullOrEmpty(value))
+                {
+                    return defaultValue;
+                }
+                else
+                {
+                    return value;
+                }
+            }
+
+            protected override ParseOptions ComputeParseOptionsWithHostValues(ParseOptions parseOptions)
```

Comment:
nit: consider making strongly typed through generics?

Reply:
Deciding no: I _could_, but there's code too that simply uses an OptionProcessor as that as the base type, and OptionProcessor<T> then wouldn't be usable without either propagating the generics or adding a base type/interface to abstract over it. And in the process, a local problem (a cast), becomes a global problem in the type system. I'm not sure that's better.

Somewhere else while doing this PR I noticed a AbstractFoo that had a derived AbstractFoo<T1, T2> which had an derived AbstractFoo<T1, T2, T3>. The hierachy could all be flattened and I got 99% there...except of the one single place that used AbstractFoo because it itself wasn't generic and needed one method. So it all unwound.